File: server/src/intTest/java/org/tctalent/server/repository/db/GetSavedListCandidatesQueryIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getCandidateOpportunity;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getCandidateSavedList;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedCandidate;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedList;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedSfJobOpp;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedUser;

import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.tctalent.server.model.db.Candidate;
import org.tctalent.server.model.db.CandidateOpportunity;
import org.tctalent.server.model.db.CandidateSavedList;
import org.tctalent.server.model.db.SalesforceJobOpp;
import org.tctalent.server.model.db.SavedList;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;
import org.tctalent.server.request.candidate.SavedListGetRequest;

public class GetSavedListCandidatesQueryIntTest extends BaseDBIntegrationTest {

  @Autowired
  private CandidateSavedListRepository repo;
  @Autowired
  private SavedListRepository savedListRepository;
  @Autowired
  private CandidateRepository candidateRepository;
  @Autowired
  private SalesforceJobOppRepository salesforceJobOppRepository;
  @Autowired
  private UserRepository userRepository;
  private SavedList testSavedList;
  private Candidate testCandidate;
  private SavedListGetRequest request;
  private GetSavedListCandidatesQuery spec;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());
    SalesforceJobOpp testSfJobOpp = getSavedSfJobOpp(salesforceJobOppRepository);
    testCandidate = getSavedCandidate(candidateRepository, getSavedUser(userRepository));

    testSavedList = getSavedList();
    testSavedList.setSfJobOpp(testSfJobOpp);
    savedListRepository.save(testSavedList);

    CandidateOpportunity co = getCandidateOpportunity();
    co.setJobOpp(testSfJobOpp);
    co.setCandidate(testCandidate);

    List<CandidateOpportunity> candidateOpportunities = new ArrayList<>();
    candidateOpportunities.add(co);
    testCandidate.setCandidateOpportunities(candidateOpportunities);
    testCandidate.addSavedList(testSavedList);
    candidateRepository.save(testCandidate);

    CandidateSavedList candidateSavedList = getCandidateSavedList(testCandidate, testSavedList);
    repo.save(candidateSavedList);

    assertNotNull(candidateSavedList.getId());

    request = new SavedListGetRequest();
  }

  @Test
  public void testCandidateSavedListsQuery() {
    spec = new GetSavedListCandidatesQuery(testSavedList, new SavedListGetRequest());
    List<Candidate> result = candidateRepository.findAll(spec);
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(1, result.size());
    assertEquals(testCandidate.getId(), result.getFirst().getId());
  }

  @Test
  public void testCandidateSavedListsQueryWithKeyword() {

    request.setKeyword(testCandidate.getCandidateNumber());
    spec = new GetSavedListCandidatesQuery(testSavedList, request);
    List<Candidate> result = candidateRepository.findAll(spec);
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(1, result.size());
    assertEquals(testCandidate.getId(), result.getFirst().getId());
  }

  @Test
  public void testCandidateSavedListsQueryWithNoMatchingKeyword() {
    request.setKeyword("NOTHING");
    spec = new GetSavedListCandidatesQuery(testSavedList, request);
    List<Candidate> result = candidateRepository.findAll(spec);
    assertNotNull(result);
    assertTrue(result.isEmpty());
  }

  @Test
  public void testCandidateSavedListsQueryWithJobOpp() {
    spec = new GetSavedListCandidatesQuery(testSavedList, new SavedListGetRequest());
    List<Candidate> result = candidateRepository.findAll(spec);
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(1, result.size());
    assertEquals(testCandidate.getId(), result.getFirst().getId());
  }

  @Test
  public void testCandidateSavedListsQueryWithClosedOpps() {
    request.setShowClosedOpps(false);
    spec = new GetSavedListCandidatesQuery(testSavedList, request);
    List<Candidate> result = candidateRepository.findAll(spec);
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(1, result.size());
    assertEquals(testCandidate.getId(), result.getFirst().getId());
  }

  @Test
  public void testCandidateSavedListsQueryWithNoMatchingCandidate() {
    SavedList nonExistentSavedList = getSavedList();
    spec = new GetSavedListCandidatesQuery(nonExistentSavedList, new SavedListGetRequest());
    List<Candidate> result = candidateRepository.findAll(spec);
    assertNotNull(result);
    assertTrue(result.isEmpty());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/GetSavedListCandidatesQueryIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.tctalent.server.model.db.Candidate
import org.tctalent.server.model.db.CandidateSavedList
import org.tctalent.server.model.db.SavedList
import org.tctalent.server.repository.db.integrationhelp.*
import org.tctalent.server.request.candidate.SavedListGetRequest

class GetSavedListCandidatesQueryIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: CandidateSavedListRepository
  @Autowired lateinit var savedListRepository: SavedListRepository
  @Autowired lateinit var candidateRepository: CandidateRepository
  @Autowired lateinit var salesforceJobOppRepository: SalesforceJobOppRepository
  @Autowired lateinit var userRepository: UserRepository

  private lateinit var testSavedList: SavedList
  private lateinit var testCandidate: Candidate
  private lateinit var candidateSavedList: CandidateSavedList

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    val testSfJobOpp = getSavedSfJobOpp(salesforceJobOppRepository)
    testCandidate = getSavedCandidate(candidateRepository, getSavedUser(userRepository))

    testSavedList = getSavedList().apply { sfJobOpp = testSfJobOpp }
    savedListRepository.save(testSavedList)

    val co =
      getCandidateOpportunity().apply {
        jobOpp = testSfJobOpp
        candidate = testCandidate
      }

    testCandidate.apply { candidateOpportunities = mutableListOf(co) }.addSavedList(testSavedList)
    candidateRepository.save(testCandidate)

    candidateSavedList = getCandidateSavedList(testCandidate, testSavedList)
    repo.save(candidateSavedList)

    assertNotNull(candidateSavedList.id)
  }

  @Test
  fun `test candidate saved lists query`() {
    val spec = GetSavedListCandidatesQuery(testSavedList, SavedListGetRequest())
    val result = candidateRepository.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isNotEmpty())
    assertEquals(1, result.size)
    assertEquals(testCandidate.id, result.first().id)
  }

  @Test
  fun `test candidate saved lists query with keyword`() {
    val request = SavedListGetRequest().apply { keyword = testCandidate.candidateNumber }
    val spec = GetSavedListCandidatesQuery(testSavedList, request)
    val result = candidateRepository.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isNotEmpty())
    assertEquals(1, result.size)
    assertEquals(testCandidate.id, result.first().id)
  }

  @Test
  fun `test candidate saved lists query with no matching keyword`() {
    val request = SavedListGetRequest().apply { keyword = "NOTHING" }
    val spec = GetSavedListCandidatesQuery(testSavedList, request)
    val result = candidateRepository.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isEmpty())
  }

  @Test
  fun `test candidate saved lists query with job opp`() {
    val spec = GetSavedListCandidatesQuery(testSavedList, SavedListGetRequest())
    val result = candidateRepository.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isNotEmpty())
    assertEquals(1, result.size)
    assertEquals(testCandidate.id, result.first().id)
  }

  @Test
  fun `test candidate saved lists query with closed opps`() {
    val request = SavedListGetRequest().apply { showClosedOpps = false }
    val spec = GetSavedListCandidatesQuery(testSavedList, request)
    val result = candidateRepository.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isNotEmpty())
    assertEquals(1, result.size)
    assertEquals(testCandidate.id, result.first().id)
  }

  @Test
  fun `test candidate saved lists query with no matching candidate`() {
    val nonExistentSavedList = getSavedList()
    val spec = GetSavedListCandidatesQuery(nonExistentSavedList, SavedListGetRequest())
    val result = candidateRepository.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isEmpty())
  }
}
